okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt

```diff
@@ -13,264 +13,287 @@
-
-      Route previousRoute = retryCurrentRoute()
-          ? transmitter.connection.route()
-          : null;
+  @Throws(IOException::class)
+  private fun findConnection(
+    connectTimeout: Int,
+    readTimeout: Int,
+    writeTimeout: Int,
+    pingIntervalMillis: Int,
+    connectionRetryEnabled: Boolean
+  ): RealConnection {
+    var foundPooledConnection = false
+    var result: RealConnection? = null
+    var selectedRoute: Route? = null
+    var releasedConnection: RealConnection?
+    val toClose: Socket?
+    synchronized(connectionPool) {
+      if (transmitter.isCanceled) throw IOException("Canceled")
+      hasStreamFailure = false // This is a fresh attempt.
+
+      val previousRoute = if (retryCurrentRoute()) {
+        // Attempt to use an already-allocated connection. We need to be careful here because our
+        transmitter.connection!!.route()
+      } else {
+        // already-allocated connection may have been restricted from creating new exchanges.
+        null
+      }
 
-      // Attempt to use an already-allocated connection. We need to be careful here because our
-      // already-allocated connection may have been restricted from creating new exchanges.
-      releasedConnection = transmitter.connection;
-      toClose = transmitter.connection != null && transmitter.connection.noNewExchanges
-          ? transmitter.releaseConnectionNoEvents()
-          : null;
+      releasedConnection = transmitter.connection
+      toClose = if (transmitter.connection != null && transmitter.connection!!.noNewExchanges) {
+        transmitter.releaseConnectionNoEvents()
+      } else {
+        null
+      }
 
       if (transmitter.connection != null) {
         // We had an already-allocated connection and it's good.
-        result = transmitter.connection;
-        releasedConnection = null;
+        result = transmitter.connection
+        releasedConnection = null
       }
 
       if (result == null) {
         // Attempt to get a connection from the pool.
         if (connectionPool.transmitterAcquirePooledConnection(address, transmitter, null, false)) {
-          foundPooledConnection = true;
-          result = transmitter.connection;
+          foundPooledConnection = true
+          result = transmitter.connection
         } else {
-          selectedRoute = previousRoute;
+          selectedRoute = previousRoute
         }
       }
     }
-    closeQuietly(toClose);
+    closeQuietly(toClose)
 
     if (releasedConnection != null) {
-      eventListener.connectionReleased(call, releasedConnection);
+      eventListener.connectionReleased(call, releasedConnection!!)
     }
     if (foundPooledConnection) {
-      eventListener.connectionAcquired(call, result);
+      eventListener.connectionAcquired(call, result!!)
     }
     if (result != null) {
       // If we found an already-allocated or pooled connection, we're done.
-      return result;
+      return result!!
```

Comment:
probably don't need those bangers; might apply elsewhere as well

Reply:
Sadly anything initialized in `synchronized` needs bangers.